src/py21cmfast/wrapper.py

```diff
@@ -2284,36 +2981,87 @@ def run_coeval(
+        generate_halobox = False
+        if flag_options.USE_HALO_FIELD and not halobox:
+            generate_halobox = True
+            halos_desc = None
+            z_halos = redshifts.copy()
+            # Photon conservaiton requires one lower redshift for the interpolation
+            if flag_options.PHOTON_CONS:
+                z_target, _, _ = get_photoncons_dz(
+                    astro_params, flag_options, z_halos[-1]
+                )
+                z_halos.append(z_target)
+                logger.info(f"calculating additional halo field at z={z_halos[-1]}")
+                # we also need a perturb field for mean boxes.
+                p = perturb_field(
+                    redshift=z_target,
+                    init_boxes=init_box,
+                    regenerate=regenerate,
+                    direc=direc,
+                    hooks=hooks,
+                )
+
+                if user_params.MINIMIZE_MEMORY:
+                    try:
+                        p.purge(force=always_purge)
+                    except OSError:
+                        pass
+                # As far as I can tell adding to the list won't affect anything else down the line
+                perturb.append(p)
+
+            pt_halos = []
+            for i, z in enumerate(z_halos[::-1]):
+                if not flag_options.FIXED_HALO_GRIDS:
+                    halos = determine_halo_list(
                         redshift=z,
                         init_boxes=init_box,
-                        user_params=user_params,
                         cosmo_params=cosmo_params,
+                        user_params=user_params,
                         astro_params=astro_params,
                         flag_options=flag_options,
-                        halo_field=determine_halo_list(
+                        regenerate=regenerate,
+                        halos_desc=halos_desc,
+                        hooks=hooks,
+                        direc=direc,
+                    )
+                    pt_halos += [
+                        perturb_halo_list(
                             redshift=z,
                             init_boxes=init_box,
+                            halo_field=halos,
+                            cosmo_params=cosmo_params,
+                            user_params=user_params,
                             astro_params=astro_params,
                             flag_options=flag_options,
                             regenerate=regenerate,
                             hooks=hooks,
                             direc=direc,
-                        ),
-                        regenerate=regenerate,
-                        hooks=hooks,
-                        direc=direc,
-                    )
-                ]
+                        )
+                    ]
+
+                    # we never want to store every halofield
+                    try:
+                        pt_halos[i].purge(force=always_purge)
```

Comment:
For now, if write==True

Reply:
Turns out `purge()` already does this, controlled by `always_purge`, which throws an error if it's not on disk and `always_purge` is False inside `_utils/OutputStruct._remove_array`. We may want to emit a warning if `write==False` and `always_purge==True`, but otherwise the behaviour is as expected.